### PR TITLE
fix: binding text to non-bindable containers and not always preferring selection

### PR DESCRIPTION
**Pull request by dwelle**
Fixes several issues around text editing and containers.

Chiefly, it's when:
1. always prefer text under cursor (with highest z-index) when editing text (multiple cases)
2. don't bind to non-text-bindable containers in specific cases when trying to edit text covered by that container

The fixes should now be covered by tests (plus added a few tests for old behavior).

The main logic to review is in `App.tsx`. All the rest is refactor, type safety, and tests.

### CODE REVIEW

File: src/components/App.tsx [hunk 1]
```diff
@@ -2009,6 +2010,7 @@ class App extends React.Component<AppProps, AppState> {
   ): NonDeleted<ExcalidrawTextElement> | null {
     const element = this.getElementAtPosition(x, y, {
       includeBoundTextElement: true,
+      preferSelected: true,
```
Comment (dwelle) [inline src/components/App.tsx:2013]: not sure if this doesn't break some sticky-note cases

Comment (dwelle) [reply src/components/App.tsx:2012]: it actually does break something: pressing enter on a selected container currently prefers to edit text that's at container center, even if the text is unbound. Unclear what the behavior should be actually, but I'll keep production behavior for now (and add a test for this).

File: src/element/textWysiwyg.test.tsx [hunk 2]
```diff
@@ -12,14 +12,210 @@ import {
...
+      expect(h.state.editingElement?.id).toBe(unboundText.id);
+    });
+
+    it("should not create bound text on ENTER if text exists at container center", () => {
+      const container = API.createElement({
+        type: "rectangle",
+        width: 100,
+      });
+      const textSize = 20;
+      const unboundText = API.createElement({
```
Comment (ad1992) [inline src/element/textWysiwyg.test.tsx:109]: why is this named `unboundText` when it has `containerID` ?

File: src/element/textWysiwyg.test.tsx [hunk 3]
```diff
@@ -12,14 +12,210 @@ import {
...
+      });
+
+      const unboundText = API.createElement({
+        type: "text",
+        text: "ola",
+        x: container.width / 2 - textSize / 2,
+        y: container.height / 2 - textSize / 2,
+        width: textSize,
+        height: textSize,
+        containerId: container.id,
```
Comment (ad1992) [inline src/element/textWysiwyg.test.tsx:153]: same for this one

File: src/components/App.tsx [hunk 4]
```diff
@@ -2156,28 +2158,43 @@ class App extends React.Component<AppProps, AppState> {
...
-    if (container) {
-      const boundTextElementId = getBoundTextElementId(container);
-      if (boundTextElementId) {
-        existingTextElement = this.scene.getElement(
-          boundTextElementId,
-        ) as ExcalidrawTextElement;
+    if (selectedElements.length === 1) {
+      if (
+        isTextElement(selectedElements[0]) &&
+        !selectedElements[0].isDeleted
```
Comment (ad1992) [inline src/components/App.tsx:2172]: is this needed ?

File: src/element/textWysiwyg.test.tsx (hunk 2 above)
Comment (dwelle) [reply src/element/textWysiwyg.test.tsx:109]: good catch

File: src/components/App.tsx (hunk 4 above)
Comment (dwelle) [reply src/components/App.tsx]: nope!

**Review by ad1992 — APPROVED** [review]

**Pull request merged**